In [35]:
import math

In [36]:
def build_sieve(p, m):
    sieve = [False] * p
    for x in range(p):
        z = (x*x - m) % p
        if z == 0:
            sieve[x] = True
        else:
            if(pow(z, (p-1)//2, p) == 1): # проверка через критерий Эйлера 
                sieve[x] = True
    return sieve

In [37]:
def overlaying_sieves(m, sieves):
    final_sieve = []
    for x in range(math.isqrt(m)+1, (m+1) // 2 + 1):
        final_sieve.append(x)
    for sieve in sieves:
        start = (math.isqrt(m)+1)%len(sieve)
        for i in range(len(final_sieve)):
            final_sieve[i] = -1 if sieve[(start+i)%len(sieve)] == False else final_sieve[i]
    return final_sieve

In [38]:
def is_perfect_square_simple(n):
    if n < 0:
        return False, None
    y = math.isqrt(n)
    if y * y == n:
        return y
    else:
        return None

In [39]:
def get_p_and_q(x, y):
    return x+y, x-y

In [40]:
def extended_gcd(a, b):  
    if b == 0:  
        return a, 1, 0  
    else:  
        gcd, x1, y1 = extended_gcd(b, a % b)  
        x = y1  
        y = x1 - (a // b) * y1  
        return gcd, x, y   


In [41]:
def p_method(m, x_1, x_2):
    step = 0
    while True:
        step += 1
        x_1 = (x_1*x_1 + 1) % m
        x_2 = (x_2*x_2 + 1) % m
        x_2 = (x_2*x_2 + 1) % m
        #print(f"{x_1} {x_2} шаг {step}")
        d = extended_gcd(abs(x_1 - x_2), m)[0]
        if d > 1:
            break
        elif d == m:
            return Exception("Не удается найти простой делитель помимо 1")
    return d, int(m/d), step

In [42]:
m = -1
result_s = None
result_p = None

while True:
    print("Что вы хотите сделать?")
    print("1 — факторизация методом квадратичного решета")
    print("2 — факторизация p-методом")
    print("3 — выйти")
    answer = input("Вводите: ").strip()
    
    if answer == "1":
        try:
            num = input("Введите m: ").strip()
            m = int(num)
            a = int(input("Введите a: ").strip())
            b = int(input("Введите b: ").strip())
            c = int(input("Введите c: ").strip())
            sieve_a = build_sieve(a, m)
            sieve_b = build_sieve(b, m)
            sieve_c = build_sieve(c, m)
            candidates = overlaying_sieves(m, [sieve_a, sieve_b, sieve_c])
            found = False
            for x in candidates:
                if(x == -1):
                    continue
                Z = x * x - m
                y = is_perfect_square_simple(Z)
                if y is not None:
                    p, q = get_p_and_q(x, y)
                    if p * q == m:
                        result_s = (p, q)
                        print(f"\nНайдены делители: p = {p}, q = {q}")
                        input("Для продолжения нажмите Enter...")
                        found = True
                        break
            if not found:
                print("\nДелители не найдены")
                result_s = None
                input("Для продолжения нажмите Enter...")
        except Exception as ex:
            print("Ошибка:", ex)
    elif answer == "2":
        try:
            if m == -1:
                num = input("Введите m: ").strip()
                m = int(num)
            x1_input = input("Введите x1 (по умолчанию 2): ").strip()
            x2_input = input("Введите x2 (по умолчанию 2): ").strip()
            x1 = int(x1_input) if x1_input else 2
            x2 = int(x2_input) if x2_input else 2
            print(f"\nВыполняем p-метод...")
            divisor, other_divisor, step = p_method(m, x1, x2)
            if divisor:
                result_p = (divisor, other_divisor, step)
                print(f"\nНайдены делители: {divisor} и {other_divisor} на шаге {step}")
                input("Для продолжения нажмите Enter...")
            else:
                print("\nДелитель не найден")
                result_p = None
                input("Для продолжения нажмите Enter...")
                
        except Exception as ex:
            print("Ошибка:", ex)
    
    elif answer == "3":
        print("Выход.")
        break
    
    else:
        print("Неверный ввод. Попробуйте снова.")

Что вы хотите сделать?
1 — факторизация методом квадратичного решета
2 — факторизация p-методом
3 — выйти
Неверный ввод. Попробуйте снова.
Что вы хотите сделать?
1 — факторизация методом квадратичного решета
2 — факторизация p-методом
3 — выйти

Найдены делители: p = 1013, q = 887
Что вы хотите сделать?
1 — факторизация методом квадратичного решета
2 — факторизация p-методом
3 — выйти

Выполняем p-метод...

Найдены делители: 1013 и 887 на шаге 16
Что вы хотите сделать?
1 — факторизация методом квадратичного решета
2 — факторизация p-методом
3 — выйти
Выход.


Для Вариант 1. 𝑚 = 898531. Получаем p = 1013, q = 887 - метод квадратичного решета. 887 и 1013 на шаге 16 - p-метод